metal/resource_metal_vlan.go

```diff
@@ -129,10 +130,35 @@ func resourceMetalVlanRead(d *schema.ResourceData, meta interface{}) error {
 
 func resourceMetalVlanDelete(d *schema.ResourceData, meta interface{}) error {
 	client := meta.(*packngo.Client)
-
-	resp, err := client.ProjectVirtualNetworks.Delete(d.Id())
+	id := d.Id()
+	vlan, resp, err := client.ProjectVirtualNetworks.Get(id, &packngo.GetOptions{Includes: []string{"instances", "instances.network_ports.virtual_networks", "internet_gateway"}})
 	if ignoreResponseErrors(httpForbidden, httpNotFound)(resp, err) != nil {
 		return friendlyError(err)
+	} else if err != nil {
+		// missing vlans are deleted
+		return nil
 	}
-	return nil
+
+	// all device ports must be unassigned before delete
```

Comment:
@t0mk we can wrap this behavior behind an attribute (`force_delete`? `unassign_on_delete`?), but if we do I think we may want it to be the default behavior.

To some extent, I see this as simply working around an API limitation.  When I ask for this type of resource to be deleted, I expect that the associations would be deleted automatically.

As we've seen in the failed tests, it seems this kind of behavior will be required to get around the dependency ordering (vlan and device are independent, port is dependent). If you try to delete a metal_vlan, it depends_on the metal_port being updated first.

We might have similar problems if a `metal_vlan` must be recreated while a `metal_port` is consuming it (or would Terraform handle that successfully?) I'll need to add a test.

Reply:
@displague I see what you are trying to do. I was thinking about this and can't come up with a scenario where user would remove a metal_vlan and wanted to keep the associations. So I think that `unassign_on_delete` is not necessary, let's keep it as you've just done it. I also don't think that the disassociation on removal will harm existing setups.

It's also OK with the vlan_attachment resources, because those will be always removed before metal_vlan (unless imported in some crazy way).

IMO, re-creation when changing will not be a problem with vlan_attachment, but will be a problem with metal_port. With metal_port, a re-created (and dissociated) VLAN will not be re-attached to the device port, because the vlan attachment happens on creation (or update) of metal_port.